### PR TITLE
Added initial Linux Cheatsheet

### PR DESCRIPTION
**Pull request by pefech**
Describes Basic Linux Usage, how to use the command line and edit files, networking and connections, pipelineing, archives and linux permissions.

### CODE REVIEW

File: source/linux/archives/data_streams.rst [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Compressing and Decompressing data streams
+==========================================
+
+Through Linux' pipes it is possible to directly compress the data stream of an programs output.
```
Comment (sriedmueller) [inline source/linux/archives/data_streams.rst]: Hi Peter,
s/of an programs/of a programs

File: source/linux/basic/filesystem.rst [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+	Note, that Linux has special paths, that are very important for the use of the file system:
+
+	* **/** is the **root directory**. In Linux **everything is represented as a file or a folder in the rootfs.** (e.g. every device can be found represented as a file in the /dev directory or programs usually in the /bin directory)
+	* **.** is a link, that exists in every directory. **It points to the directory, that it is in**.
+	* **..** is a link, that exists in every directory. **It points to the parent directory of the directory it is in**. (e.g. /home/user/.. points to /home)
+	* **~** is everywhere useable. **It points to the home directory of the current user**
+
+
+	Note, that a path can be an absolute or relative. An absolute path starts every time with the root file system (e.g. /home/user/documents). A relative path is relative to the working directory and can even be only a file name (e.g. documents, ../../home)
```
Comment (sriedmueller) [inline source/linux/basic/filesystem.rst]: Maybe use always instead of every time.

File: source/linux/basic/filesystem.rst [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		Outputs the current working directory
+	Example Usage:
+		.. code-block:: bash
+
+			host$ pwd
+
+ls [path]
+^^^^^^^^^
+	Full Name:
+		**L** ist **D** irectory
```
Comment (sriedmueller) [inline source/linux/basic/filesystem.rst]: Doesn't it have to be **L**i**S**t directory ?

File: source/linux/basic/filesystem.rst [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^
+	Full Name:  
+		**L** i **n** k
+
+	Description:
+		Links files or folders
+	Example Usage:
+		.. code-block:: bash
+
+			host$ host$ ln flash.bin flash_link.bin
```
Comment (sriedmueller) [inline source/linux/basic/filesystem.rst]: There are to host$ here,
Also above you mention the -s option but your example does not use it. Maybe you can add a description for symbolic links.

File: source/linux/basic/usersystem.rst [hunk 5]
```diff
@@ -0,0 +1,70 @@
+Usersytem Basics
+================
+
+Linux has a multiuser user system which allows multiple users to work on one system at the same time. 
+Linux' user system includes a permission system (more on that on another Page). The following commands will give you a basic overview:
```
Comment (sriedmueller) [inline source/linux/basic/usersystem.rst]: Maybe you can add a link here?

File: source/linux/editing_files/vi.rst [hunk 6]
```diff
@@ -0,0 +1,62 @@
+Vi
```
Comment (sriedmueller) [inline source/linux/editing_files/vi.rst:3]: Maybe you can reference the tutorial vim includes.

**Review by sriedmueller — COMMENTED** [review]
Hi Peter, its not a complete review, but a few comments I had when skimming through your changes.

File: source/linux/archives/list_of_archives.rst [hunk 7]
```diff
@@ -0,0 +1,91 @@
+List of Archives
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/archives/list_of_archives.rst]: We should also add 7z, since it is also used often.

File: source/linux/basic/filesystem.rst [hunk 8]
```diff
@@ -0,0 +1,188 @@
+Filesystem
+==========
+
+The Linux filesystem is build like every other filesystem (e.g. Windows or macOS).It's build out of folders, that can contain subfolders, files or links to other files.
+A Linux path can look like this: "/home/user/documents". 
+When using a console, the console is running in a folder, the working directory. You can make changes on the working directory, files or folders with this commands:
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/filesystem.rst]: Maybe "command line" is a more convenient term.

File: source/linux/basic/filesystem.rst [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+	Note, that a path can be an absolute or relative. An absolute path starts every time with the root file system (e.g. /home/user/documents). A relative path is relative to the working directory and can even be only a file name (e.g. documents, ../../home)
+
+
+.. hint::
+	Note, that the notation for command arguments is as follows:
+
+	- arguments enclosed with <> are placeholders for required arguments.
+	- arguments enclosed with [] are placeholders for optional arguments.
+
+cd
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/filesystem.rst]: `cd <path>`

File: source/linux/basic/filesystem.rst [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+	Description:
+		Changes current working directory
+
+	Example Usage:
+		.. code-block:: bash
+
+			host$ cd /var/lib
+
+		changes working directory to the lib folder that is in the var folder which is in the / folder
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/filesystem.rst]: we could replace / folder with "root directory"

File: source/linux/basic/filesystem.rst [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+		changes working directory to the documents folder, that is in the current working directory
+
+pwd
+^^^
+	Full Name:
+		**P** rint **W** orking **D** irectory
+
+	Description:
+		Outputs the current working directory
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/filesystem.rst]: Outputs the path to the current working directory

File: source/linux/basic/filesystem.rst [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+		**L** ist **D** irectory
+
+	Description:
+		Lists the files and folders in the working directory or in the given path
+	Example Usage:
+		.. code-block:: bash
+
+			host$ ls documents/projects
+		
+		lists the contents of the project folder, thats in the documents folder which is in the working directory
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/filesystem.rst:76]: You could just write something like:
 lists the contents of folder ./documents/projects/

File: source/linux/basic/filesystem.rst [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+	Full Name:  
+		**M** ma **k** e **Dir** ectory
+
+	Description:
+		Creates a Directory
+	Example Usage:
+		.. code-block:: bash
+
+			host$ mkdir documents/projects/newProject
+
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/filesystem.rst:89]: A comment would be good

File: source/linux/basic/filesystem.rst [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+	Full Name:  
+		**M** o **v** e
+
+	Description:
+		Moves files or folders
+	Example Usage:
+		.. code-block:: bash
+
+			host$ mv flash.bin output/flash.bin
+
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/filesystem.rst:116]: A comment would be good.

You could add something that moving a file/folder is similar to renaming it.

File: source/linux/basic/filesystem.rst [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+	Description:
+		Copies files or folders
+	Example Usage:
+		.. code-block:: bash
+
+			host$ cp flash.bin flash_copy.bin
+		
+		Creates a exact copy of flash.bin to the new file flash_copy.bin
+
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/filesystem.rst:165]: you could add a example to copy a wohle folder

File: source/linux/basic/filesystem.rst [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+	Description:
+		Searches a file in the given path, including all subdirectories
+	Example Usage:
+		.. code-block:: bash
+
+			host$ find . -name "flash.bin"
+
+		Searches the file "flash.bin" in the working directory
+		
+dd
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/filesystem.rst]: Add parameters

File: source/linux/basic/installing_uninstalling_software.rst [hunk 17]
```diff
@@ -0,0 +1,74 @@
+Installing and Uninstalling Software
+====================================
+
+Most Linux distributions come with a package manager like apt or yum. It's made to simplify the process of downloading and installing new software and its dependencies on a users system. 
+In the following only debian based systems are covered because the VM you get from PHYTEC by purchasing a Kit is a debian based Distribution.
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/installing_uninstalling_software.rst]: Maybe more like: 
....
The following sections only cover Debian based systems, since the PHYTEC Development Environment, which ships with all PHYTEC kits, uses a virtual machine (VM) with a Debian based Linux distribution.
...

File: source/linux/basic/installing_uninstalling_software.rst [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+
+apt install <packages>
+^^^^^^^^^^^^^^^^^^^^^^
+    Full Name
+        **A** dvanced **P** ackaging **T** ool
+
+    Description
+        Installs a package with its dependencies.
+
+        Has to be executed with super user privileges.
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/installing_uninstalling_software.rst]: Add a cross reference to privileges / sudo

File: source/linux/basic/other.rst [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+
+    Description
+        Lists all active Processes with their needed resources in a table. It has functions to kill Processes or to filter the list.
+
+    Example Usage
+        .. code-block:: bash
+
+            host$ htop
+        
+minicom -b <baudrate> -P <port>
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/other.rst:37]: Maybe we could add a separate section like "Working with serial interfaces", where minicom is described and also the basic usage of serial ttys with cat and echo?

File: source/linux/basic/usersystem.rst [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+        Executes a Command with superuser privileges.
+        **Warning:** only execute a command with superuser privileges, if you know exactly what you are doing!
+
+        Sometimes you will notice, that you cannot execute a command, because you does not have all permissions. 
+        If you execute it with sudo, you should have all needed permissions. (e.g. passwd has to be run with sudo)
+	
+    Example Usage
+        .. code-block:: bash
+
+            host$ sudo passwd demouser
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/usersystem.rst]: Maybe we should use a other example, like apt, since passwd can also be used without super user privileges, as long as it's your own user.

File: source/linux/basic/wildcards.rst [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+        The Shell replaces the asterisk with everything, that is available.
+        In the case of the example the shell would replace the \*.o with a list of all files that fit the wildcard. 
+        But only in the working directory. The asterisk does not include subdirectories.
+    
+    Usage
+        For Example deleting all files with .o extension
+        .. code-block:: bash
+            
+            $host rm \*.o
+
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/basic/wildcards.rst]: You could also describe ** for subdirs:
like `ll **/*.o` for listing all .o files in all subdirs.

File: source/linux/editing_files/vi.rst [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
++------------------+------------------+
+| Mode             | Key              |
++==================+==================+
+| INSERT           | i                |
++------------------+------------------+
+| VISUAL           | v                |
++------------------+------------------+
+
+INSERT mode
+    In INSERT mode you can actively edit the text of the file you have loaded. You can use the cursor keys to navigate and edit as normal as you would in an Office tool.
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/editing_files/vi.rst]: ... office tool.

File: source/linux/editing_files/vi.rst [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+    | p                | Paste            |
+    +------------------+------------------+
+    | <line> SHIFT-G   | Go to line       |
+    +------------------+------------------+
+    | gg               | Go to first line |
+    +------------------+------------------+
+    | G                | Go to last line  |
+    +------------------+------------------+
+    | /<phrase>        | Search for phrase|
+    +------------------+------------------+
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/editing_files/vi.rst:88]: You could also describe the multiplayer like:
`<number> yy` to copy <number> lines 

and force quit would be useful: 
`q!`

File: source/linux/networking/connecting.rst [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+    
+    Usage:
+        .. code-block:: bash
+                
+                host$ sftp [user@]host
+    
+    Example:
+        .. code-block:: bash
+                
+                host$ user@10.0.0.137
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/networking/connecting.rst]: sftp command is missing

File: source/linux/networking/network_config.rst [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+
+    Usage:
+        ping [options] [host]
+
+    Example:
+        .. code-block:: bash
+
+            host$ ping <address>
+
+ip addr
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/networking/network_config.rst]: Some more information about ip would be nice, since it will replace ifconfig in the most distros

File: source/linux/permissions/reading.rst [hunk 26]
```diff
@@ -0,0 +1,37 @@
+Reading Permissions
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/permissions/reading.rst]: This could cause some misunderstandings, something like `How to read permissions` could be better.

File: source/linux/permissions/permissions.rst [hunk 27]
```diff
@@ -0,0 +1,16 @@
+Permissions
+***********
+
+.. warning::
+
+    Linux permissions is a very complex topic. This cheatsheet only covers the basics
+
+Overview
+========
+
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/permissions/permissions.rst]: We should also describe what groups are and how to edit them. 
Also root / superuser concept should be described.

File: source/linux/pipes_grep/pipes.rst [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+
+    The output of a program can even redirected into a File.
+    This is done with the ">>" or the ">" operator.
+
+    The >> operator appends a file with the output.
+    The > operator replaces the file with the output. Example:
+
+    .. code-block:: bash
+
+        host$ sudo dmesg >> dmesg.log
```
Comment (EverythingElseWasAlreadyTaken) [inline source/linux/pipes_grep/pipes.rst:32]: you sould also describe the dmesg command somewhere, if you show it here.

**Review by EverythingElseWasAlreadyTaken — COMMENTED** [review]
I noted some minor issues that may need to be discussed.
Overall, good work. :)

**Comment by EverythingElseWasAlreadyTaken** [comment]
Could you also please rebase your branch to the current main branch and test the pdf generation?
I've seen some problems with the code blocks in the wildcard chapter, this could be caused by a wrong indentation.

File: source/linux/basic/other.rst (hunk 19 above)
Comment (pefech) [reply source/linux/basic/other.rst:37]: I think we should add this to the ToDo issue and do this another time.

File: source/linux/editing_files/vi.rst (hunk 6 above)
Comment (pefech) [reply source/linux/editing_files/vi.rst:3]: I'm sorry, but I don't know exactly what you mean.

**Comment by pefech** [comment]
Fixed everything commented or commented back for further discussion.

File: source/linux/archives/data_streams.rst (hunk 1 above)
Comment (EverythingElseWasAlreadyTaken) [reply source/linux/archives/data_streams.rst]: @pefech please fix this

File: source/linux/basic/other.rst (hunk 19 above)
Comment (EverythingElseWasAlreadyTaken) [reply source/linux/basic/other.rst:37]: Okay, I'll add this to the ToDo Issue

File: source/linux/editing_files/vi.rst (hunk 6 above)
Comment (EverythingElseWasAlreadyTaken) [reply source/linux/editing_files/vi.rst:3]: There is a vim tutorial program, which ships with the most linux distros. 
Just type `vimtutor` in your terminal.

**Review by EverythingElseWasAlreadyTaken — COMMENTED** [review]
Looks good to me :)

File: source/linux/networking/connecting.rst [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+        downloading from remote:
+            .. code-block:: bash
+
+                host$ scp user@example.com:~/doc.txt .
+
+
+        uploading to remote:
+            .. code-block:: bash
+
+                host$ scp ./firmware.hex user@example.com:~/
```
Comment (tremmet) [inline source/linux/networking/connecting.rst]: Use IP of our target

File: source/linux/networking/connecting.rst [hunk 30]
```diff
@@ -0,0 +1,58 @@
...
+        .. code-block:: bash
+
+            host$ scp [<user>@<host>:]<source> [<user>@<host>:]<destination>
+
+    Example:
+
+        downloading from remote:
+            .. code-block:: bash
+
+                host$ scp user@example.com:~/doc.txt .
```
Comment (tremmet) [inline source/linux/networking/connecting.rst]: Use IP of our target.

File: source/linux/networking/connecting.rst [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+
+    Usage:
+        .. code-block:: bash
+
+            host$ ssh [user@]host
+
+    Example:
+        .. code-block:: bash
+
+            host$ ssh user@10.0.0.137
```
Comment (tremmet) [inline source/linux/networking/connecting.rst]: We should use our standard IP address to connect to our target here: 192.168.3.11 
And not any random address.

File: source/linux/networking/connecting.rst [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+ssh
+^^^
+    Description:
+        Starts a **s** ecure **s** hell sessions with the given host and user.
+        When the connection is established, you are working on the remote system.
+
+    Usage:
+        .. code-block:: bash
+
+            host$ ssh [user@]host
```
Comment (tremmet) [inline source/linux/networking/connecting.rst]: Maybe replace the host in [user@]host with some other name. Like destination ?

File: source/linux/editing_files/nano.rst [hunk 33]
```diff
@@ -0,0 +1,17 @@
+Nano
```
Comment (tremmet) [inline source/linux/editing_files/nano.rst:1]: Nano is not installed on default on our targets. We should at least add a note about this.

File: source/linux/basic/usersystem.rst [hunk 34]
```diff
@@ -0,0 +1,82 @@
+Usersytem Basics
+================
+
+Linux has a multiuser user system which allows multiple users to work on one system at the same time. 
+Linux' user system includes a permission system (more on that here: :ref:`permissions` ).
+The following commands will give you a basic overview:
```
Comment (tremmet) [inline source/linux/basic/usersystem.rst]: Same with this chapter here. This is actually out of scope. The customer might even think this is how this is supposed to be handled on the target. I am not sure if this is not too confusing.

File: source/linux/basic/usersystem.rst [hunk 35]
```diff
@@ -0,0 +1,82 @@
+Usersytem Basics
```
Comment (tremmet) [inline source/linux/basic/usersystem.rst]: s/Usersystem/User system/g

File: source/linux/basic/installing_uninstalling_software.rst [hunk 36]
```diff
@@ -0,0 +1,76 @@
+Installing and Uninstalling Software
+====================================
```
Comment (tremmet) [inline source/linux/basic/installing_uninstalling_software.rst]: I am not sure if this is really useful. There are so many distributions out there. For my perspective we should concentrate on the things that are relevant in interaction with our boards. This is not.

File: source/linux/networking/connecting.rst [hunk 37]
```diff
@@ -0,0 +1,58 @@
...
+
+                host$ scp user@example.com:~/doc.txt .
+
+
+        uploading to remote:
+            .. code-block:: bash
+
+                host$ scp ./firmware.hex user@example.com:~/
+
+sftp
```
Comment (tremmet) [inline source/linux/networking/connecting.rst:44]: sftp seems not supported on our targets. I do not know if this is then worth mentioning.

File: source/linux/networking/network_config.rst [hunk 38]
```diff
@@ -0,0 +1,86 @@
+Chaning network config
+======================
+
+ifconfig
```
Comment (tremmet) [inline source/linux/networking/network_config.rst]: We should not explain ifconfig. It is deprecated!

File: source/linux/permissions/important_commands.rst [hunk 39]
```diff
@@ -0,0 +1,48 @@
...
+        **user** **mod** ification
+    
+    example:
+        .. code-block:: bash
+        
+            usermod -a -G <group> <user>
+
+        adds the user to the specified group.
+    
+adduser
```
Comment (tremmet) [inline source/linux/permissions/important_commands.rst]: This had been explained already above.

**Review by tremmet — CHANGES_REQUESTED** [review]

File: source/linux/networking/connecting.rst (hunk 37 above)
Comment (EverythingElseWasAlreadyTaken) [reply source/linux/networking/connecting.rst:44]: I've added a note about it.

File: source/linux/editing_files/nano.rst (hunk 33 above)
Comment (EverythingElseWasAlreadyTaken) [reply source/linux/editing_files/nano.rst:1]: I've added a note about it.

File: source/linux/basic/installing_uninstalling_software.rst (hunk 36 above)
Comment (EverythingElseWasAlreadyTaken) [reply source/linux/basic/installing_uninstalling_software.rst]: The idea behind this was to also support the phytec development environment, which we provide for our customers.
I'll extend this text a little, to make it more clear.

File: source/linux/basic/usersystem.rst (hunk 34 above)
Comment (EverythingElseWasAlreadyTaken) [reply source/linux/basic/usersystem.rst]: Also here, I'll extend the text a little to make it more clear that this is just for desktop distros.

Comment (EverythingElseWasAlreadyTaken) [reply source/linux/basic/usersystem.rst]: I've added a note and moved this chapter to the users and permissions chapter. 
It would be good if you could take a quick look at it.

File: source/linux/basic/installing_uninstalling_software.rst (hunk 36 above)
Comment (EverythingElseWasAlreadyTaken) [reply source/linux/basic/installing_uninstalling_software.rst]: I've added a note and a link to our Yocto manual.

**Review by tremmet — APPROVED** [review]

**Review by NineTec — APPROVED** [review]
